### PR TITLE
Introducing the maximum size preview

### PR DESCRIPTION
**Pull request by oparoz**
### Fixes
- Always store the preview returned by Imagemagick before resizing it. https://github.com/owncloud/core/issues/12465
### The problem

You've just uploaded a 100MB TIFF file that you need to share with a restricted audience. Here is what currently happens once the file has been uploaded.
- A 36x36 thumbnail is generated in 1 minute
- Go to the Gallery, an album thumbnail might be generated in 1 minute
- Go inside the album, another thumbnail will be generated in 1 minute
- Click on the picture to get a full screen preview. 1 minute
- Go back to the Files app and share the file
- Client 1 visits the link. 1 minute to get the preview
- Client 2 with a different screen resolution visits the link. 1 minute again

You're starting to get it. Every time a slightly different preview size is asked, a full conversion of that 100MB TIFF to PNG is performed. This is highly inefficient.
### Proposed solution
- Enforce maximum preview dimensions in the config. They're currently set to null.
- The first time we're asked to generate a preview we'll generate one of the maximum dimension indicated in the configuration and all future resizing requests will be done on that preview in order to not waste time converting the same large file over and over

Default size is 2048x2048
The maximum size preview is keeping the aspect ratio of the original and is not upscaled.
It's is immediately stored in the cache with a special `max` tag and size identifiers based on its real size.

Every time a preview of a different size is asked, we store it in the cache. This also speeds up operation at the cost of disk space.

This make a huge difference in libraries holding several large files, for people using apps such as the Gallery.

Recreated after the loss of #13582

### CODE REVIEW

**Comment by karlitschek** [comment]
@georgehrke What is your opinion?

File: lib/private/preview.php [hunk 1]
```diff
@@ -82,9 +82,9 @@ public function __construct($user = '', $root = '/', $file = '', $maxX = 1, $max
 		$this->userView = new \OC\Files\View('/' . $user);
 
 		//set config
-		$this->configMaxX = \OC_Config::getValue('preview_max_x', null);
-		$this->configMaxY = \OC_Config::getValue('preview_max_y', null);
-		$this->maxScaleFactor = \OC_Config::getValue('preview_max_scale_factor', 2);
+		$this->configMaxX = \OC::$server->getConfig()->getSystemValue('preview_max_x', 2048);
```
Comment (georgehrke) [inline lib/private/preview.php]: Can we maybe modify the constructor to expect some `OCP\IConfig` and if null is given fall back to `\OC::$server->getConfig()`?

File: lib/private/preview.php [hunk 2]
```diff
@@ -396,6 +398,12 @@ private function isCachedBigger($fileId) {
 			}
```
Comment (georgehrke) [inline lib/private/preview.php]: please add `$width === 'max' || ...` to the if condition

File: lib/private/preview.php (hunk 1 above)
Comment (oparoz) [inline lib/private/preview.php]: Seems doable, but is it the right patch to modify the signature? Wouldn't it be better to introduce that in a PR which can be given as a reference to the change so that people can start modifying their code? @MorrisJobke @LukasReschke

Comment (georgehrke) [inline lib/private/preview.php]: People should never create a \OC\Preview instance directly and use `getPreviewManager` anyway.
(but removing all direct creations of \OC\Preview in core should definitely be done in a separate pr. (will take care))

**Comment by georgehrke** [comment]
Code looks good to me otherwise

File: lib/private/preview.php (hunk 1 above)
Comment (oparoz) [inline lib/private/preview.php]: Agreed, except that, sometimes, it's not possible to use it as PreviewManager is not yet a true replacement of \OC\Preview
See https://github.com/owncloud/core/issues/12772 for what's missing

But if it's confirmed that the signature change should happen here, I'll do it.

File: lib/private/preview.php [hunk 3]
```diff
@@ -921,7 +1056,7 @@ public static function post_delete($args, $prefix='') {
 	 * @return bool
 	 */
 	public static function isAvailable(\OC\Files\FileInfo $file) {
-		if (!\OC_Config::getValue('enable_previews', true)) {
+		if (!\OC::$server->getConfig()->getSystemValue('enable_previews', true)) {
```
Comment (nickvergessen) [inline lib/private/preview.php]: This will conflict with https://github.com/owncloud/core/pull/14857

File: lib/private/preview.php (hunk 1 above)
Comment (nickvergessen) [inline lib/private/preview.php]: That's the right approach, but will conflict greatly with https://github.com/owncloud/core/pull/14857

Maybe you can wait, until that is done, since it is a major rework on the same file aswell ;)

Comment (oparoz) [inline lib/private/preview.php]: Sounds reasonable. I'll let you adjust the signature then and will update this PR when you're done.

**Comment by oparoz** [comment]
Since there were no newly introduced public methods, I just did an extensive test.

I'm also forcing a resize since the image provider is not doing it's resizing job properly and since that's what people are going to use the most.

**Comment by nickvergessen** [comment]
Needs to be rebased

**Comment by oparoz** [comment]
@georgehrke @MorrisJobke @PVince81 @DeepDiver1975 
Could me merge this before the PR goes, once again, out of sync?

**Comment by oparoz** [comment]
Maybe @icewind1991 and @jancborchardt want to test it as well since this benefits the Gallery apps.

File: lib/private/preview.php [hunk 4]
```diff
@@ -698,11 +703,161 @@ private function resizeAndCrop() {
...
+						'core', 'Generating preview for "' . $file . '" with "' . get_class($provider)
+						. '"', \OCP\Util::DEBUG
+					);
+
+					/** @var $provider Provider */
+					$preview = $provider->getThumbnail(
+						$file, $this->configMaxX, $this->configMaxY, $scalingUp = false, $this->fileView
+					);
+
+					if (!($preview instanceof \OC_Image)) {
```
Comment (nickvergessen) [inline lib/private/preview.php]: Use the interface please

Comment (oparoz) [inline lib/private/preview.php]: Yep, thanks.

**Comment by MorrisJobke** [comment]
Tested and works :+1:

**Comment by georgehrke** [comment]
mentioned issues fixed, works :+1:
